src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEc2Response()
+        {
+            var response = AwsHttpClientHelper.GetEC2Response().GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            string sessionToken = null;
+            if (properties != null)
+            {
+                foreach (var pair in properties)
+                {
+                    switch (pair.Key.ToUpperInvariant())
+                    {
+                        case "AWS_SESSION_TOKEN":
+                            sessionToken = (string)pair.Value;
+                            break;
+                        default:
+                            throw new ArgumentException($"Unknown AWS property '{pair.Key}'.", nameof(properties));
+                    }
+                }
+            }
+
+            return sessionToken;
+        }
+
+        private static SecureString ToSecureString(string str)
+        {
+            if (str == null)
+            {
+                return null;
+            }
+
+            var secureString = new SecureString();
+            foreach (var c in str)
+            {
+                secureString.AppendChar(c);
+            }
+            secureString.MakeReadOnly();
+
+            return secureString;
+        }
+
+        private static void ValidateAwsCredentials(AwsCredentials awsCredentials)
+        {
+            if (awsCredentials.Username == null && (awsCredentials.Password != null || awsCredentials.SessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (awsCredentials.Username != null && awsCredentials.Password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
```

Comment:
I didn't understand what is being validated here.

Reply:
It sounds like the validated rules are described in here https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#obtaining-credentials
But I'm not sure about this statement:
>In other words, regardless of how Drivers obtain credentials the only valid combination of credentials is an access key ID and a secret access key or an access key ID, a secret access key and a session token.

Cat it be translated like: 
>( access key ID && a secret access key) || (an access key ID && a secret access key && a session token)
?